damnit/backend/sandboxing.py

```diff
@@ -0,0 +1,149 @@
+        ]
+
+        self.command_binds: list[tuple[str, str, str]] = []
+
+        for path in (
+            "/bin",
+            "/etc/resolv.conf",
+            "/lib",
+            "/lib64",
+            "/sbin",
+            "/usr",
+        ):
+            self.add_bind(Path(path), ro=True)
+
+        if Path("/gpfs/exfel/sw/software").exists():
+            self.add_bind(Path("/gpfs/exfel/sw/software"), ro=True)
+
+    def add_bind(
+        self, source: Path, dest: Path | None = None, ro: bool = False
+    ) -> None:
+        """Adds a bind mount to the sandbox.
+
+        !!! warning
+
+            Bind mounts are done on inodes, if a program updates a file mounted into the
+            sandbox by overwriting it then the inode changes and the file will not be
+            in sync between host/sandbox. To avoid this mount the parent directory.
+
+        Args:
+            source (Path): The source path to be bind mounted.
+            dest (Path, optional): The destination path in the sandbox. If not provided, the source path is used.
+            ro (bool, optional): Whether the bind mount should be read-only. Defaults to False.
+
+        Raises:
+            ValueError: If the source path is not absolute.
+        """
+        if not source.is_absolute():
+            raise ValueError("Source path must be absolute")
+
+        if dest is None:
+            dest = source
+
+        self.command_binds.append(
+            (
+                f"--{'ro-' if ro else ''}bind",
+                shlex.quote(str(source)),
+                shlex.quote(str(dest)),
+            )
+        )
+
+    def add_bind_proposal(self, proposal_id: int) -> None:
+        """Adds bind mounts for a proposal directory and its contents.
+
+        Args:
+            proposal_id (int): The ID of the proposal.
+
+        Raises:
+            FileNotFoundError: If the proposal directory is not found.
+        """
+        proposal_dir = Path(find_proposal(f"p{proposal_id:06d}"))
+
+        self.add_bind(proposal_dir)
+
+        for path in proposal_dir.iterdir():
+            self.add_bind(path.resolve())
+
+    def add_bind_venv(self, python_exec: Path) -> None:
+        """Adds all paths required by a virtual environment to the sandbox.
+
+        This function will use the given python executable to first call `sys.prefix` to
+        check if the executable is in a venv, if it is then `sysconfig.get_paths()` is
+        used to find required paths and add them paths as read-only binds.
+
+        Args:
+            python_exec (Path): The path to the Python executable.
+
+        Raises:
+            subprocess.CalledProcessError: If the command to get the virtual environment paths fails.
+        """
+        venv = subprocess.check_output(
+            [python_exec, "-c", "import sys; print(sys.prefix != sys.base_prefix)"]
+        ).decode("utf-8")
+
+        if venv == "False":
+            return
```

Comment:
I think it would be surprising that this method is a complete no-op if the Python you point to isn't a venv. Especially if people use conda envs - they're *technically* not venvs, but I think people usually expect things to work in the same way.

Would it make sense to turn this into `add_bind_python` and try to do the right thing with the target Python, venv or not?

It could also start by adding the main env directory (i.e. the bit before `bin/python`) and only add extra paths if they're not under that; it should be the same, but a shorter command is easier to make sense of if we ever need to.

(Also, I think you'd actually end up with `"False\n"` here if the target is not a venv, so the check would fail. It might be worth using JSON to send details back, just to avoid fiddly details like this.)